src/exporter/http_server.cc

```diff
@@ -0,0 +1,168 @@
+#include <boost/beast/http.hpp>
+#include <boost/beast/version.hpp>
+#include <boost/thread/thread.hpp>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+
+namespace beast = boost::beast;   // from <boost/beast.hpp>
+namespace http = beast::http;     // from <boost/beast/http.hpp>
+namespace net = boost::asio;      // from <boost/asio.hpp>
+using tcp = boost::asio::ip::tcp; // from <boost/asio/ip/tcp.hpp>
+
+class http_connection : public std::enable_shared_from_this<http_connection> {
+public:
+  http_connection(tcp::socket socket) : socket_(std::move(socket)) {}
+
+  // Initiate the asynchronous operations associated with the connection.
+  void start() {
+    read_request();
+    check_deadline();
+  }
+
+private:
+  tcp::socket socket_;
+  beast::flat_buffer buffer_{8192};
+  http::request<http::dynamic_body> request_;
+  http::response<http::dynamic_body> response_;
+
+  net::steady_timer deadline_{socket_.get_executor(), std::chrono::seconds(60)};
+
+  // Asynchronously receive a complete request message.
+  void read_request() {
+    // auto self = shared_from_this();
+
+    http::async_read(socket_, buffer_, request_,
+                     [&](beast::error_code ec, std::size_t bytes_transferred) {
+                       boost::ignore_unused(bytes_transferred);
+                       if (!ec)
+                         process_request();
+                     });
+  }
+
+  // Determine what needs to be done with the request message.
+  void process_request() {
+    response_.version(request_.version());
+    response_.keep_alive(request_.keep_alive());
+
+    switch (request_.method()) {
+    case http::verb::get:
+      response_.result(http::status::ok);
+      create_response();
+      break;
+
+    default:
+      // We return responses indicating an error if
+      // we do not recognize the request method.
+      response_.result(http::status::bad_request);
+      response_.set(http::field::content_type, "text/plain");
+      beast::ostream(response_.body())
+          << "Invalid request-method '" << std::string(request_.method_string())
+          << "'";
+      break;
+    }
+
+    write_response();
+  }
+
+  // Construct a response message based on the program state.
+  void create_response() {
+    if (request_.target() == "/") {
+      response_.set(http::field::content_type, "text/html; charset=utf-8");
+      beast::ostream(response_.body())
+          << "<html>\n"
+          << "<head><title>Ceph Exporter</title></head>\n"
+          << "<body>\n"
+          << "<h1>Ceph Exporter</h1>\n"
+
+          << "<p><a href='/metrics'>Metrics</a></p>"
+          << "</body>\n"
+          << "</html>\n";
+    } else if (request_.target() == "/metrics") {
+      response_.set(http::field::content_type, "text/plain; charset=utf-8");
+      DaemonMetricCollector &collector = collector_instance();
+      std::string metrics = collector.get_metrics();
```

Comment:
```c++
int main(int argc, char** argv) {
  boost::thread server_thread(http_server_thread_entrypoint);
  DaemonMetricCollector &collector = collector_instance();
  collector.main();
  server_thread.join();
}
```
with `http_server` and `DaemonMetricCollector` running in separate threads, don't we need to protect this access under a mutex?

Reply:
Which access? The only thing I see that could need a mutex is DaemonMetricCollector::get_metrics

Reply:
right. doesn't that need a mutex?

Reply:
the http thread is reading from `DaemonMetricCollector::metrics`, and the collector thread is writing to it. that's a data race, so it either needs to be fixed with atomics or locks

`std::string` isn't atomic, so both the reads and the writes to `DaemonMetricCollector::metrics` need to be protected by a mutex